File: 78-Subsets.cpp
//78-Subsets.cpp

#include<iostream>
#include<vector>
using namespace std;

//2017/03/22
//Time:O(2^N)
//Space:O(1)
//遍历+迭代
class Soluiton{
public:
    vector<vector<int>> subsets(vector<int>& nums)
    {
        vector<vector<int>> r={{}};
        for(int k=0;k<nums.size();k++){
        	int end=r.size();
            for(int i=0;i<end;i++){
            	vector<int> t=r[i];
            	t.push_back(nums[k]);
                r.push_back(t);
            }
        }

        return r;
    }

};

int main()
{
    Soluiton s;
    vector<int> v={1,2,5,3,6};
    vector<vector<int>> r=s.subsets(v);

    for(int k=0;k<r.size();k++){
        for(int j=0;j<r[k].size();j++)
            cout<<r[k][j]<<" ";
        cout<<endl;
    }

    return 0;
}

File: 78. Subsets.py
# -*- coding: utf-8 -*-
# @Time    : 2019/2/12
# @Author  : qirui
# @FileName: 78. Subsets.py

# reference: https://leetcode.com/problems/combination-sum/discuss/16502/A-general-approach-to-backtracking-questions-in-Java-(Subsets-Permutations-Combination-Sum-Palindrome-Partitioning)
# Backtrack/DFS (general)
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        def backtrack(tmp, start):
            res.append(tmp)
            for i in range(start, N):
                backtrack(tmp + [nums[i]], i + 1)

        N = len(nums)
        res = []
        backtrack([], 0)
        return res


if __name__ == '__main__':
    S = Solution()
    print(S.subsets([1, 2, 3]))
